### PR TITLE
Part 7: Working with form

### PR DESCRIPTION
**Pull request by hiepn-1339**
Part 7: Working with form

### CODE REVIEW

**Comment by hiepn-1339** [comment]
Eslint:

image

**Comment by hiepn-1339** [comment]
ready

File: src/controllers/author.controller.ts [hunk 1]
```diff
@@ -3,6 +3,10 @@ import asyncHandler from 'express-async-handler';
 import * as authorService from '../services/author.service';
 import { t } from 'i18next';
 import { Author } from '../entities/author.entity';
+import {validationResult, Result, body} from 'express-validator';
```
Comment (huongnt-2545) [inline src/controllers/author.controller.ts:6]:
```suggestion
import { validationResult, Result, body } from 'express-validator';
```

File: src/controllers/author.controller.ts [hunk 2]
```diff
@@ -30,10 +34,81 @@ export const authorDetail = [checkExistsAuthor, (req: IAuthorRequest, res: Respo
...
+
+      if (newAuthor) {
+        res.redirect(`/author/${newAuthor.id}`);
+      } else {
+        res.redirect('/author');
+      }
+    }
+
+    res.render('authors/form', {
+      title: 'Create Author',
```
Comment (huongnt-2545) [inline src/controllers/author.controller.ts:59]: I18n text

File: src/middlewares/author.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  return true;
+});
+
+export const checkValidDateOfDeath = body('dateOfDeath').custom((value, { req }) => {
+  if (!value || !req.body.dateOfBirth) return true;
+
+  const dateOfDeath = new Date(value);
+  const dateOfBirth = new Date(req.body.dateOfBirth);
+
+  if (dateOfBirth > dateOfDeath) {
```
Comment (huongnt-2545) [inline src/middlewares/author.middleware.ts:23]: thêm case nếu dateofdeath > current_date nữa e nhé

File: src/middlewares/author.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { body } from 'express-validator';
+import { t } from 'i18next';
+
+export const checkValidDateOfBirth = body('dateOfBirth').custom((value, { req }) => {
+  if (!value || !req.body.dateOfDeath) return true;
+
+  const dateOfBirth = new Date(value);
+  const dateOfDeath = new Date(req.body.dateOfDeath);
+
+  if (dateOfBirth > dateOfDeath) {
```
Comment (huongnt-2545) [inline src/middlewares/author.middleware.ts:10]: thêm case nếu dateofbirth > current_date

File: src/services/author.service.ts [hunk 5]
```diff
@@ -16,3 +16,25 @@ export const authorDetail = async (id: number) => {
   });
 };
 
+export const authorCreate = async (data: any) => {
+  const author = new Author();
+  author.firstName = data.firstName;
+  author.familyName = data.familyName;
+  author.dateOfBirth = data.dateOfBirth ? data.dateOfBirth : null;
+  author.dateOfDeath = data.dateOfDeath ? data.dateOfDeath : null;
```
Comment (huongnt-2545) [inline src/services/author.service.ts:24]: dùng method `create`

File: src/services/book.service.ts [hunk 6]
```diff
@@ -35,3 +36,39 @@ export const bookDetail = async (id: number) => {
     where: { id: id },
   });
 };
+
+export const bookCreateGet = async () => {
+  return await Promise.all([
+    authorRepository.find({}),
```
Comment (huongnt-2545) [inline src/services/book.service.ts:42]: thêm order author nhé

**Comment by hiepn-1339** [comment]
Eslint:

image

**Comment by hiepn-1339** [comment]
ready

**Review by huongnt-2545 — APPROVED** [review]
merged

**Pull request merged**